### PR TITLE
feat(geo): add bulk geofence upload

### PR DESCRIPTION
**Pull request by AaronZyLee**
#### Description of changes

Add a new subcommand `populate` for geo category, which enables developer to upload their own GeoJSON file to the existing geofence collection
#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  } else {
+    //Non-first should be clockwise
+    if (!isClockWise) {
+      throw new Error('The non-frist linear ring is interior and should be closewise.')
+    }
+  }
+};
+
+//Check the wind direction of linear ring. Assume the linear ring is valid and closed
+//Refer https://stackoverflow.com/questions/1165647/how-to-determine-if-a-list-of-polygon-points-are-in-clockwise-order
```
Comment (phani-srikar) [inline packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts:76]: nit: remove this link please

File: packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  //Check if first position is identical to last one
+  if (!(linearRing[0][0] === linearRing[numPoint-1][0] && linearRing[0][1] === linearRing[numPoint-1][1])) {
+    throw new Error(`Linear ring "${linearRing}" should have the identical values for first and last position.`);
+  }
+  //Check polygon wind direction
+  const isClockWise: boolean = isClockWiseLinearRing(linearRing);
+  if (isFirstRing) {
+    //First Ring should be counter clockwise
+    if (isClockWise) {
+      throw new Error('The first linear ring is exterior ring and should be counter-closewise.')
```
Comment (phani-srikar) [inline packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts:65]: nit: clockwise

File: packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  } else {
+    //Non-first should be clockwise
+    if (!isClockWise) {
+      throw new Error('The non-frist linear ring is interior and should be closewise.')
```
Comment (phani-srikar) [inline packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts:70]: Please re-phrase this to `The interior linear ring should be clockwise.`

File: packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * Validate linear ring for those not captured by JSON Schema, incluidng first and last identical position, winding direction
+ * @param linearRing Array of coordinates
+ * @param isFirstRing Whether it is first(exterior) ring
+ */
+const validateLinearRing = (linearRing: Array<Array<number>>, isFirstRing: boolean) => {
+  const numPoint = linearRing.length;
+  //Check if first position is identical to last one
+  if (!(linearRing[0][0] === linearRing[numPoint-1][0] && linearRing[0][1] === linearRing[numPoint-1][1])) {
+    throw new Error(`Linear ring "${linearRing}" should have the identical values for first and last position.`);
```
Comment (phani-srikar) [inline packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts:58]: can we pass in the Geofence ID and throw a more useful error like `Linear ring "${linearRing}" should have the identical values for first and last position for the Geofence with ID ${geofenceID}`

**Review by phani-srikar — CHANGES_REQUESTED** [review]

**Review by phani-srikar — APPROVED** [review]
LGTM!

File: packages/amplify-category-geo/src/provider-controllers/populate.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  })
+  return {
+    CollectionName: populateParam.collectionToPopulate,
+    Entries
+  }
+}
+
+const bulkUploadGeofence = async (params: GeofenceCollectionParams, region: string) => {
+  const service = new Location({region});
+  return await service.batchPutGeofence(params).promise();
```
Comment (Amplifiyer) [inline packages/amplify-category-geo/src/provider-controllers/populate.ts:96]: Is this going to upload all the geofences in the file at once? There seems to be a limit of 10 geofences per batch https://docs.aws.amazon.com/location-geofences/latest/APIReference/API_BatchPutGeofence.html#API_BatchPutGeofence_RequestBody

File: packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
+import { readFileSync } from 'fs-extra';
+import Ajv from 'ajv';
+import { FeatureCollection, IdentifierOption } from './populateParams';
+import GeoJSONSchema from 'amplify-category-geo/schema/GeoJSONSchema.json';
+import { v4 as uuid } from 'uuid';
+import { printer } from 'amplify-prompts';
+
+
+export const validateGeoJSONFile = (geoJSONFilePath: string, uniqueIdentifier: string = 'id', identifierOption: IdentifierOption = IdentifierOption.RootLevelID) => {
```
Comment (Amplifiyer) [inline packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts:9]: There is another limit of 1000 number of coordinates/vertices in the polygon https://docs.aws.amazon.com/location-geofences/latest/APIReference/API_BatchPutGeofenceRequestEntry.html#API_BatchPutGeofenceRequestEntry_Contents

**Review by Amplifiyer — COMMENTED** [review]
Looks good. Left some small comments.

File: packages/amplify-category-geo/src/provider-controllers/populate.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+export const populateResource = async (context: $TSContext) => {
+  const geofenceCollectionResources = ((await context.amplify.getResourceStatus()).allResources as any[])
+  .filter(resource => resource.service === ServiceName.GeofenceCollection);
+  if (geofenceCollectionResources.length === 0) {
+    throw new Error('Geofence collection is not found. Use `amplify geo add` to create a new geofence collection.')
+  }
+  //Get provisioned geofence collection name
+  const collectionNames = geofenceCollectionResources.map(collection => {
+    if (!collection.output || !collection.output.Name || !collection.output.Region) {
+      throw new Error(`Geofence ${collection.resourceName} is not provisioned yet.`)
```
[the remark this answers is not in the document]
Comment (AaronZyLee) [reply packages/amplify-category-geo/src/provider-controllers/populate.ts:20]: Yes. Will add the fix in the msg

File: packages/amplify-category-geo/src/provider-controllers/populate.ts (hunk 5 above)
Comment (AaronZyLee) [reply packages/amplify-category-geo/src/provider-controllers/populate.ts:115]: Added the limit for each batch

File: packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts (hunk 6 above)
Comment (AaronZyLee) [reply packages/amplify-category-geo/src/service-utils/validateGeoJSONFile.ts:10]: Added the vertices number check

**Review by Amplifiyer — APPROVED** [review]
Looks good. If possible add unit tests for the new logic related to batches and validations.

**Pull request merged**